dirdiff: init at 2.1

khumba opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This is a Tcl/Tk directory diff and merge tool that is simple to use, and is something I've been missing on NixOS.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
  - This new package takes 160K according to `du -sh`.  It's closure size is 49MiB.
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

khumba committed:
dirdiff: init at 2.1

r-rmcgibbo commented:
Result of `nixpkgs-review pr 113177` at 285367ec run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/c69b7b3902d2ddb37751f155ec7691cf">dirdiff</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/text/dirdiff/default.nix:23:3:
```
   |
23 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

r-rmcgibbo commented:
Result of `nixpkgs-review pr 113177` at 285367ec run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/184ed8be0aaf4b036ceb6981030c6c0b">dirdiff</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/text/dirdiff/default.nix:23:3:
```
   |
23 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

fabaff commented:
I would like to suggest to create a `desktopItem` as it's a GUI tool. Not a blocker though.

The last release was made over 15 years ago. Very interesting that it's still compiling :smile:

khumba commented:
> I would like to suggest to create a `desktopItem` as it's a GUI tool. Not a blocker though.

Yes, this should exist.  Added, thanks!

> The last release was made over 15 years ago. Very interesting that it's still compiling :smile:

C and Tcl software sure has good longevity :).

Changes and lint addressed, and I also switched to using `installFlags` rather than a custom `installPhase`.

SuperSandro2000 commented on pkgs/tools/text/dirdiff/default.nix:
```diff
@@ -0,0 +1,55 @@
+    })
+  ];
+
+  meta = {
```
```suggestion
  meta = with lib; {
```

SuperSandro2000 commented on pkgs/tools/text/dirdiff/default.nix:
```diff
@@ -0,0 +1,55 @@
+      directory trees and for merging changes from one tree into another.
+    '';
+    homepage = "https://www.samba.org/ftp/paulus/";
+    license = lib.licenses.gpl2Plus;
+    maintainers = with lib.maintainers; [ khumba ];
+    platforms = lib.platforms.linux;
```
```suggestion
    license = licenses.gpl2Plus;
    maintainers = with maintainers; [ khumba ];
    platforms = platforms.linux;
```

SuperSandro2000 commented on pkgs/tools/text/dirdiff/default.nix:
```diff
@@ -0,0 +1,55 @@
+  preInstall = ''
+    mkdir -p $out/bin $out/lib
+  '';
+  installFlags = [ "BINDIR=\${out}/bin" "LIBDIR=\${out}/lib" ];
```
```suggestion
  installFlags = [ "BINDIR=${placeholder "out"}/bin" "LIBDIR=${placeholder "out"}/lib" ];
```

khumba replied:
Interesting, I hadn't seen this magic function, thanks!  In grepping Nixpkgs it seems like `$(out)` is the preferred syntax for interpolating into `installFlags`, do you mind if I use that?

SuperSandro2000 replied:
I would prefer if you use placeholder. Greping nixpkgs is not a good metric because it also counts all the dust and rusted packages.

khumba replied:
Okay, I see that this is new norm, done.  I wish the Nix(pkgs?) manual documented when these should be used; I admit I'm not totally sure why variable substitution isn't a good idea here :).

SuperSandro2000 commented on pkgs/tools/text/dirdiff/default.nix:
```diff
@@ -0,0 +1,55 @@
+{ fetchurl, lib, stdenv, tcl, tk
+, copyDesktopItems, makeDesktopItem
+}:
+stdenv.mkDerivation {
```
```suggestion
}:

stdenv.mkDerivation {
```

khumba commented on pkgs/tools/text/dirdiff/default.nix:
```diff
@@ -0,0 +1,39 @@
+  buildInputs = [ tcl tk ];
+
+  # Some light path patching.
+  inherit tcl tk;
```
These environment variables are used by `substituteInPlace --subst-var`.  I could use `--subst-var-by` instead to avoid the environment variables...

SuperSandro2000 replied:
I am not sure if you are using this correct as I have never seen this before.

khumba replied:
This just adds `tcl` and `tk` variables to the environment.  `--subst-var-by` seems to be preferred for dependencies though, so I've switched it to use that.  (`--subst-var` appears only used for output paths.)

khumba commented:
Feedback addressed, thanks @SuperSandro2000.

SuperSandro2000 commented on pkgs/tools/text/dirdiff/default.nix:
```diff
@@ -0,0 +1,62 @@
+{ copyDesktopItems, fetchurl, lib, makeDesktopItem, stdenv, tcl, tk }:
+
+let
+  pname = "dirdiff";
+  version = "2.1";
+in
+
+stdenv.mkDerivation {
+  inherit pname version;
```
```suggestion
stdenv.mkDerivation rec {
  pname = "dirdiff";
  version = "2.1";
```

khumba replied:
Done.
